### PR TITLE
Lazily build a case-insensitive comparison for performance

### PR DESCRIPTION
**Pull request by 333fred**
This linear search was responsible for 5% of CPU time in a customer trace. This is one possible approach to reducing the time we spend doing the search, while keeping the standard case-sensitive comparison the same. This path should only be hit by error recovery in VB, so it shouldn't have an impact on C# or non-error scenarios.

### CODE REVIEW

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (ToddGrun) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3652]: >_lazyForwardedTypesToAssemblyIndexMap [](http://example.com/codeflow?start=49&length=37)

The majority of entries have zero entries (at least in Roslyn). Maybe it would be worthwhile to have an empty dictionary to use for those instead of allocating a new one?

Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3652]: Fair observation.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (ToddGrun) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3650]: >OneOrMany [](http://example.com/codeflow?start=64&length=9)

Instead of keeping a OneOrMany, couldn't we just keep the first one? (the caller only uses values[0])

**Comment by ToddGrun** [comment]
>                 var typesToAssemblyIndexMap = new Dictionary<string, (int FirstIndex, int SecondIndex)>();

nit: you probably already considered it, but the allocation of this could be moved inside the loop after determining there is something to add

---
Refers to: src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3689 in f896148. [](commit_id = f8961482f6e36590b47b04026790d0b51ae9d87f, deletion_comment = False)

**Comment by 333fred** [comment]
Test Insertion: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/494697

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for reviews

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (ToddGrun) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3661]: >_lazyForwardedTypesToAssemblyIndexMap [](http://example.com/codeflow?start=49&length=37)

I assume ordering doesn't matter in the case of matching keys?

Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3661]: Correct. The existing code preferred whatever came first in the dictionary, which is pretty much undefined since it isn't an ordered dictionary in any way.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (sharwell) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:50]: 💡 Would be good to explicitly name the tuple elements

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:50]: >s_sharedEmptyCaseInsensitiveForwardedTypes [](http://example.com/codeflow?start=63&length=42)

Can we share ```s_sharedEmptyForwardedTypes``` field for this purpose? It doesn't look like the instances have different characteristics.

**Comment by AlekseyTs** [comment]
> This path should only be hit by error recovery

It looks like this statement is not accurate. An attempt to lookup a forwarded type is a normal way of operation during success scenarios as well.

>  in VB, so it shouldn't have an impact on C# or non-error scenarios.

This looks accurate though

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3632]: >matchedName = fullName [](http://example.com/codeflow?start=20&length=22)

It looks like there is a behavior change here, we are no longer returning the matched casing.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3631]: >out matchedName [](http://example.com/codeflow?start=105&length=15)

We probably should add a test to cover this aspect (return of the actual spelling)

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3668]: >_lazyCaseInsensitiveForwardedTypesToAssemblyIndexMap = caseInsensitiveMap.ToImmutable(); [](http://example.com/codeflow?start=16&length=88)

Since ```ImmutableSegmentedDictionary``` is a struct, the assignment is no longer an atomic operation, and, therefore, is not thread-safe.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3631]: Will try and devise a test that hits this codepath.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3668]: Thanks for the catch.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: >_lazyForwardedTypesToAssemblyIndexMap = ImmutableSegmentedDictionary<string, (int FirstIndex, int SecondIndex)>.Empty; [](http://example.com/codeflow?start=20&length=118)

It looks like the assignment is no longer thread-safe.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: >_lazyForwardedTypesToAssemblyIndexMap = typesToAssemblyIndexMap.ToImmutable(); [](http://example.com/codeflow?start=20&length=78)

It looks like the assignment is no longer thread-safe.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3631]: It looks like it is not possible to observe this. https://github.com/dotnet/roslyn/blob/main/src/Compilers/VisualBasic/Portable/Binding/Binder_Symbols.vb#L421 is the only codepath that calls this function with `ignoreCase=True`, and it does not use the `ByRef metadataName` parameter again after the difference would have been noticed.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 6)

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3631]: There is an existing test I would have expected to observe the difference, if any difference were able to be observed: https://github.com/dotnet/roslyn/blob/main/src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/TypeForwarders.vb#L745-L782

**Comment by 333fred** [comment]
@AlekseyTs I think I've addressed all your feedback.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6), assuming CI is passing

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3657]: Do we also need InterlockedInitialize here?

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (ToddGrun) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3631]: >TryGetKey [](http://example.com/codeflow?start=85&length=9)

Isn't TryGetKey O(n)?

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3632]: >matchedName = fullName [](http://example.com/codeflow?start=20&length=22)

Since ```matchedName``` parameter is meaningless now, let's remove it and follow up at the call sites accordingly.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 8)

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (ToddGrun) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3632]: > Since `matchedName` parameter is meaningless now, let's remove it and follow up at the call sites accordingly.

Seems reasonable, @333fred?

Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs]: @ToddGrun Aleksey and I were working offline on determining that we were wrong about the observability of this change 🙂.

File: src/Compilers/Core/Portable/MetadataReader/PEModule.cs
Comment (ToddGrun) [inline src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3634]: Would it be preferable to just return the tuple including the string here? Seems like it might be cleaner than having a return value with two items in it and an out parameter (and also would prevent an allocation, right?)

Comment (ToddGrun) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3634]: I guess the normal case of getting it from _lazyForwardedTypesToAssemblyIndexMap would require an allocation then (unless it was changed to include the original string too).

Comment (333fred) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3634]: No allocation prevented. `ValueTuple` is a value type. I'm going to leave the code as is.

Comment (ToddGrun) [reply src/Compilers/Core/Portable/MetadataReader/PEModule.cs:3634]: Oh, for some reason I was thinking the (...) syntax created a Tuple, not a ValueTuple. Whoops!

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 9)

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by 333fred** [comment]
@RikkiGibson @dotnet/roslyn-compiler for a second review.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**
